### DISCUSSION TITLE
Epic React update-deps script

### DISCUSSION THREAD

**Discussion by kentcdodds** [johnlindquist/kit#125] [Share]
Whenever I update [react-workshop-app](https://github.com/kentcdodds/react-workshop-app), I need to update all my EpicReact workshop repos. I have a script in each one called `update-deps` which looks like this:

https://github.com/kentcdodds/react-hooks/blob/119a8c9182664e7b28e6a426b7ca861f045f729e/scripts/update-deps

So I've put together a little script that opens an individual terminal window for each of the repos and runs the update-deps script and pushes. Really straightforward and saves me a lot of time.

```javascript
// Menu: Update EpicReact deps
// Description: Update all the dependencies in the epicreact workshop repos

const repos = [
  'advanced-react-hooks',
  'advanced-react-patterns',
  'bookshelf',
  'react-fundamentals',
  'react-hooks',
  'react-performance',
  'react-suspense',
  'testing-react-apps',
]

const script = `git add -A && git stash && git checkout main && git pull && ./scripts/update-deps && git commit -am "update all deps" --no-verify && git push && git status`
for (const repo of repos) {
  const scriptString = JSON.stringify(
    `cd ~/code/epic-react/${repo} && ${script}`,
  )
  exec(
    `osascript -e 'tell application "Terminal" to activate' -e 'tell application "Terminal" to do script ${scriptString}'`,
  )
}
```

**Comment by johnlindquist** [comment]
Awesome, very cool scenario to automate away!

There is a built-in [`terminal` helper](https://github.com/johnlindquist/kit/blob/0d5ccc042678dd927dd2d8cb218ae4c393dbea2d/src/preload/os/mac.ts#L26) which does exactly what you're doing with exec:

```js
await terminal(scriptString)
```

should work. There's also an [`iterm` helper](https://github.com/johnlindquist/kit/blob/0d5ccc042678dd927dd2d8cb218ae4c393dbea2d/src/preload/os/mac.ts#L36) if that's more your jam.

**Reply by kentcdodds** [reply]
Oh cool. I noticed that the iterm helper tries to open a new tab rather than a new window if possible. Would love to have that in the terminal helper as well!

Also, when I tried to update this to use the terminal helper I got this in the `console.log` file:

```
[2021-04-12 13:21:10.198] [info] 42:44: syntax error: Expected end of line but found identifier. (-2741)
```

My file doesn't even have a line 42 so I'm not sure where this is coming from 🤔

**Reply by johnlindquist** [reply]
Can you please post the script using the terminal helper? Just want to be 100% sure I'm on the same page as you.

Re: tabs, if it's possible in AppleScript, then it's possible. I'll hit the Google and see what I can find

**Reply by kentcdodds** [reply]
Yep, here it is:

```js
// Menu: Update EpicReact deps
// Description: Update all the dependencies in the epicreact workshop repos

const repos = [
  'advanced-react-hooks',
  'advanced-react-patterns',
  'bookshelf',
  'react-fundamentals',
  'react-hooks',
  'react-performance',
  'react-suspense',
  'testing-react-apps',
]

const script = `git add -A && git stash && git checkout main && git pull && ./scripts/update-deps && git commit -am "update all deps" --no-verify && git push && git status`
for (const repo of repos) {
  const scriptString = JSON.stringify(
    `cd ~/code/epic-react/${repo} && ${script}`,
  )
  terminal(scriptString) // (I don't need to await this because I don't care about waiting)
}
```

**Reply by johnlindquist** [reply]
Thanks, found it:

> Note: A temporary fix is to use single quotes for `'update all deps'`

### The bug:

`scriptString` is passed to:
1. The `terminal` function, which builds another string
2. Then to the internal `applescript` function which escapes the string passed to osascript

So the `scriptString` was escaped too early AND I wasn't handling double quotes properly in the `terminal` function.

This will be fixed in the next build. Cheers.
